File: backend/api/v1/experiences.py
from fastapi import (
    APIRouter
)

from pydantic import (
    BaseModel
)

from backend.services.experience_service import (
    ExperienceService
)


router = APIRouter()

service = ExperienceService()


class ExperienceCreateRequest(
    BaseModel
):

    robot_id: str

    task: str

    strategy: str

    reward: float


@router.post(
    "/experiences"
)
def create_experience(
    request:
    ExperienceCreateRequest
):

    experience = (
        service.create_experience(
            request.robot_id,
            request.task,
            request.strategy,
            request.reward
        )
    )

    return {
        "id":
            experience.id,
        "robot_id":
            experience.robot_id,
        "task":
            experience.task,
        "strategy":
            experience.strategy,
        "reward":
            experience.reward
    }


@router.get(
    "/experiences"
)
def get_experiences():

    data = (
        service.get_experiences()
    )

    return {
        "count": len(data),
        "experiences": [
            {
                "id": item.id,
                "robot_id":
                    item.robot_id,
                "task":
                    item.task,
                "strategy":
                    item.strategy,
                "reward":
                    item.reward
            }
            for item in data
        ]
    }


@router.get(
    "/experiences/{robot_id}"
)
def get_robot_experiences(
    robot_id: str
):

    data = (
        service.get_robot_experiences(
            robot_id
        )
    )

    return {
        "robot_id":
            robot_id,
        "count":
            len(data),
        "experiences": [
            {
                "id": item.id,
                "task": item.task,
                "strategy":
                    item.strategy,
                "reward":
                    item.reward
            }
            for item in data
        ]
    }

File: frontend/js/experiences.js
async function loadExperiences() {

    const data =
        await getData(
            "/experiences"
        );

    renderExperiences(
        data.experiences
    );

    document.getElementById(
        "experience-count"
    ).textContent =
        data.count;
}

function renderExperiences(
    experiences
) {

    const body =
        document.getElementById(
            "experience-body"
        );

    body.innerHTML = "";

    experiences.forEach(
        exp => {

            body.innerHTML += `
            <tr class="border-b border-slate-800">

                <td class="py-3">${exp.id}</td>

                <td class="py-3">
                    ${exp.robot_id}
                </td>

                <td class="py-3">
                    ${exp.task}
                </td>

                <td class="py-3">
                    ${exp.strategy}
                </td>

                <td class="py-3">
                    ${exp.reward}
                </td>

            </tr>
            `;
        }
    );
}

async function createExperience() {

    const robotId =
        document.getElementById(
            "robot-id"
        ).value;

    const task =
        document.getElementById(
            "task"
        ).value;

    const strategy =
        document.getElementById(
            "strategy"
        ).value;

    const reward =
        parseFloat(
            document.getElementById(
                "reward"
            ).value
        );

    await fetch(
        "http://127.0.0.1:8000/api/v1/experiences",
        {
            method: "POST",

            headers: {
                "Content-Type":
                    "application/json"
            },

            body: JSON.stringify({

                robot_id: robotId,

                task: task,

                strategy: strategy,

                reward: reward
            })
        }
    );

    await loadExperiences();
}

async function filterRobot() {

    const robotId =
        document.getElementById(
            "robot-filter"
        ).value;

    if (!robotId) {

        await loadExperiences();

        return;
    }

    const data =
        await getData(
            `/experiences/${robotId}`
        );

    renderExperiences(
        data.experiences
    );
}

loadExperiences();